File: ts/tests/local/test-marketplace.mjs
/**
 * Marketplace resource test — runs against prodv2 gateway.
 * Usage: node test-marketplace.mjs
 */

import { ImbraceClient } from '@imbrace/sdk'

const ACCESS_TOKEN = process.env.IMBRACE_ACCESS_TOKEN || 'acc_c8c27f3b-e147-4735-b641-61e8d3706692'
const GATEWAY      = process.env.IMBRACE_GATEWAY_URL  || 'https://app-gatewayv2.imbrace.co'

const client = new ImbraceClient({ accessToken: ACCESS_TOKEN, gateway: GATEWAY })
const marketplace = client.marketplace

let passed = 0, failed = 0, skipped = 0

function ok(label, detail = '') {
  console.log(`  ✓ ${label}${detail ? `  →  ${String(detail).slice(0, 80)}` : ''}`)
  passed++
}

function fail(label, err) {
  console.error(`  ✗ ${label}: ${err?.message ?? JSON.stringify(err)}`)
  failed++
}

function skip(label, reason) {
  console.log(`  - ${label}  (skipped: ${reason})`)
  skipped++
}

// ── [1] List use case templates ───────────────────────────────────────────────

console.log('\n[1] Use case templates')
try {
  const res = await marketplace.listUseCaseTemplates()
  const list = Array.isArray(res) ? res : (res.data ?? [])
  ok('listUseCaseTemplates()', `${list.length} templates`)
} catch (e) { fail('listUseCaseTemplates()', e) }

// ── [2] List products ─────────────────────────────────────────────────────────

console.log('\n[2] Products')
let firstProductId = null
try {
  const res = await marketplace.listProducts({ limit: 3 })
  const list = Array.isArray(res) ? res : (res.data ?? [])
  ok('listProducts()', `${list.length} products`)
  if (list.length > 0) firstProductId = list[0]._id ?? list[0].id
} catch (e) { fail('listProducts()', e) }

// ── [3] Get product ───────────────────────────────────────────────────────────

console.log('\n[3] Get product')
if (!firstProductId) {
  skip('getProduct()', 'no products found')
} else {
  try {
    const p = await marketplace.getProduct(firstProductId)
    ok('getProduct()', `id=${p._id ?? p.id} name=${p.name}`)
  } catch (e) { fail('getProduct()', e) }
}

// ── [4] List orders ───────────────────────────────────────────────────────────

console.log('\n[4] Orders')
let firstOrderId = null
try {
  const res = await marketplace.listOrders({ limit: 3 })
  const list = Array.isArray(res) ? res : (res.data ?? [])
  ok('listOrders()', `${list.length} orders`)
  if (list.length > 0) firstOrderId = list[0]._id ?? list[0].id
} catch (e) {
  if (e?.message?.includes('404')) skip('listOrders()', 'orders route not available on this env')
  else fail('listOrders()', e)
}

// ── [5] Get order ─────────────────────────────────────────────────────────────

console.log('\n[5] Get order')
if (!firstOrderId) {
  skip('getOrder()', 'no orders found')
} else {
  try {
    const o = await marketplace.getOrder(firstOrderId)
    ok('getOrder()', `id=${o._id ?? o.id}`)
  } catch (e) { fail('getOrder()', e) }
}

// ── [6] List email templates ──────────────────────────────────────────────────

console.log('\n[6] Email templates')
try {
  const res = await marketplace.listEmailTemplates()
  const list = Array.isArray(res) ? res : (res.data ?? [])
  ok('listEmailTemplates()', `${list.length} templates`)
} catch (e) {
  if (e?.message?.includes('400') || e?.message?.includes('404')) skip('listEmailTemplates()', 'requires org context or route unavailable')
  else fail('listEmailTemplates()', e)
}

// ── [7] List categories ───────────────────────────────────────────────────────

console.log('\n[7] Categories')
try {
  const res = await marketplace.listCategories()
  const list = Array.isArray(res) ? res : (res.data ?? [])
  ok('listCategories()', `${list.length} categories`)
} catch (e) {
  if (e?.message?.includes('404')) skip('listCategories()', 'categories route not available on this env')
  else fail('listCategories()', e)
}

// ── Summary ───────────────────────────────────────────────────────────────────

console.log(`\n${'─'.repeat(50)}`)
console.log(`  ${passed} passed  |  ${failed} failed  |  ${skipped} skipped`)
if (failed > 0) process.exit(1)

File: py/tests/local/test_marketplace.py
"""
Marketplace resource test — sync + async, runs against prodv2 gateway.
Usage:
    pip install -e ".[dev]"
    python tests/local/test_marketplace.py
"""
import os
import asyncio
from imbrace import ImbraceClient
from imbrace.async_client import AsyncImbraceClient

ACCESS_TOKEN = os.environ.get("IMBRACE_ACCESS_TOKEN", "acc_c8c27f3b-e147-4735-b641-61e8d3706692")
GATEWAY      = os.environ.get("IMBRACE_GATEWAY_URL",  "https://app-gatewayv2.imbrace.co")

passed = 0
failed = 0
skipped = 0

def ok(label, detail=""):
    global passed
    detail_str = f"  →  {str(detail)[:80]}" if detail else ""
    print(f"  ✓ {label}{detail_str}")
    passed += 1

def fail(label, err):
    global failed
    print(f"  ✗ {label}: {err}")
    failed += 1

def skip(label, reason):
    global skipped
    print(f"  - {label}  (skipped: {reason})")
    skipped += 1

# ─────────────────────────────────────────────────────────────────────────────
# Sync
# ─────────────────────────────────────────────────────────────────────────────

def test_sync():
    print("\n--- Synchronous Tests ---")
    client = ImbraceClient(access_token=ACCESS_TOKEN, gateway=GATEWAY)
    marketplace = client.marketplace

    # [1] Use case templates
    print("\n[1] Use case templates")
    try:
        res = marketplace.list_use_case_templates()
        lst = res.get("data", res) if isinstance(res, dict) else res
        ok("list_use_case_templates()", f"{len(lst) if isinstance(lst, list) else '?'} templates")
    except Exception as e: fail("list_use_case_templates()", e)

    # [2] Products
    print("\n[2] Products")
    first_product_id = None
    try:
        res = marketplace.list_products(limit=3)
        lst = res.get("data", res) if isinstance(res, dict) else res
        ok("list_products()", f"{len(lst) if isinstance(lst, list) else '?'} products")
        if isinstance(lst, list) and len(lst) > 0:
            first_product_id = lst[0].get("_id") or lst[0].get("id")
    except Exception as e: fail("list_products()", e)

    # [3] Get product
    print("\n[3] Get product")
    if not first_product_id:
        skip("get_product()", "no products found")
    else:
        try:
            p = marketplace.get_product(first_product_id)
            p_data = p.get("data", p) if isinstance(p, dict) else p
            ok("get_product()", f"id={p_data.get('_id') or p_data.get('id')} name={p_data.get('name')}")
        except Exception as e: fail("get_product()", e)

    # [4] Orders
    print("\n[4] Orders")
    first_order_id = None
    try:
        res = marketplace.list_orders(limit=3)
        lst = res.get("data", res) if isinstance(res, dict) else res
        ok("list_orders()", f"{len(lst) if isinstance(lst, list) else '?'} orders")
        if isinstance(lst, list) and len(lst) > 0:
            first_order_id = lst[0].get("_id") or lst[0].get("id")
    except Exception as e:
        if "404" in str(e): skip("list_orders()", "route not available")
        else: fail("list_orders()", e)

    # [5] Get order
    print("\n[5] Get order")
    if not first_order_id:
        skip("get_order()", "no orders found")
    else:
        try:
            o = marketplace.get_order(first_order_id)
            o_data = o.get("data", o) if isinstance(o, dict) else o
            ok("get_order()", f"id={o_data.get('_id') or o_data.get('id')}")
        except Exception as e: fail("get_order()", e)

    # [6] Email templates
    print("\n[6] Email templates")
    try:
        res = marketplace.list_email_templates()
        lst = res.get("data", res) if isinstance(res, dict) else res
        ok("list_email_templates()", f"{len(lst) if isinstance(lst, list) else '?'} templates")
    except Exception as e:
        if "400" in str(e) or "404" in str(e): skip("list_email_templates()", "requires org context or route unavailable")
        else: fail("list_email_templates()", e)

    # [7] Categories
    print("\n[7] Categories")
    try:
        res = marketplace.list_categories()
        lst = res.get("data", res) if isinstance(res, dict) else res
        ok("list_categories()", f"{len(lst) if isinstance(lst, list) else '?'} categories")
    except Exception as e:
        if "404" in str(e): skip("list_categories()", "route not available")
        else: fail("list_categories()", e)

# ─────────────────────────────────────────────────────────────────────────────
# Async
# ─────────────────────────────────────────────────────────────────────────────

async def test_async():
    print("\n--- Asynchronous Tests ---")
    async with AsyncImbraceClient(access_token=ACCESS_TOKEN, gateway=GATEWAY) as client:
        marketplace = client.marketplace
        try:
            res = await marketplace.list_use_case_templates()
            lst = res.get("data", res) if isinstance(res, dict) else res
            ok("async list_use_case_templates()", f"count={len(lst) if isinstance(lst, list) else '?'}")
        except Exception as e: fail("async list_use_case_templates()", e)

if __name__ == "__main__":
    test_sync()
    asyncio.run(test_async())

    print(f"\n{'─' * 50}")
    print(f"  {passed} passed  |  {failed} failed  |  {skipped} skipped")
    if failed > 0:
        raise SystemExit(1)
